### PR TITLE
Federation e2e support for AWS

### PR DESCRIPTION
**Pull request by colhom**
I've observed e2e test failures on the two local runs I did, but the framework seems to come up successfully. [logs](http://pastebin.com/tsJpKUc4). Ideas on this?

I'm in the process of validating GKE as well, and will modify the title if it succeeds.

\cc @nikhiljindal @quinton-hoole 

[Analytics]()

### CODE REVIEW

File: federation/manifests/federation-apiserver-deployment.yaml [hunk 1]
```diff
@@ -22,7 +22,11 @@ spec:
         - --etcd-servers=http://localhost:2379
         - --service-cluster-ip-range={{.FEDERATION_SERVICE_CIDR}}
         - --secure-port=443
+        {{if eq .KUBERNETES_PROVIDER "aws"}}
+        - --external-hostname={{.FEDERATION_API_HOST}}
```
Comment (nikhiljindal) [inline federation/manifests/federation-apiserver-deployment.yaml]: I am not sure why is this AWS specific.

**Comment by nikhiljindal** [comment]
ok to test

@kubernetes/sig-cluster-federation

**Comment by nikhiljindal** [comment]
Looking at the logs, federation apiserver test passed, and the 2 federation service tests failed. Those 2 tests are failing on GCE as well right now. federation apiserver test passing means that bring up worked fine. So this is good!

**Comment by quinton-hoole** [comment]
Awesome!  Thanks for this @colhom. I'll take a deeper look into the logs to confirm that the reasons for those two tests failures are the same reasons for failures on GCE and GKE. If so, we already have fixes in the works.

@madhusudancs @matchstick FYI

**Comment by quinton-hoole** [comment]
@colhom FYI manual GKE validation was done by Kelsey Hightower recently.  See:

https://github.com/kelseyhightower/kubernetes-cluster-federation
https://youtu.be/86jZdmAjWns

**Comment by quinton-hoole** [comment]
@colhom You seem to be running as an old version of the tests, as the line numbers in your log file don't line up with the line numbers in upstream/master:HEAD.

Could you perhaps rebase and rerun?

Note also that you're probably running into https://github.com/kubernetes/kubernetes/pull/27802 which is now merged.

So all-in-all, a rebase and rebuild would be very useful right now.

File: federation/cluster/common.sh [hunk 2]
```diff
@@ -135,6 +135,30 @@ function create-federation-api-objects {
 	    fi
 	    sleep 5
 	done
+
+	# Poll until DNS record for federation-apiserver is propagated
```
Comment (nikhiljindal) [inline federation/cluster/common.sh]: What happens if we do not wait?

File: federation/cluster/common.sh [hunk 3]
```diff
@@ -135,6 +135,30 @@ function create-federation-api-objects {
...
+	# TODO(colhom): REQUIREMENT for any other providers which use a DNS name for routing to loadbalancers
+	# right now it's just AWS that does this.
+	which nslookup
+	set +o errexit
+	if [[ "$KUBERNETES_PROVIDER" == "aws" ]];then
```
Comment (nikhiljindal) [inline federation/cluster/common.sh]: I am fine with removing this if and doing this polling for all cloud providers

**Comment by quinton-hoole** [comment]
@colhom You definitely also want to rebase to get https://github.com/kubernetes/kubernetes/pull/27794
Pretty sure you'll run into that one too.

File: federation/manifests/federation-apiserver-deployment.yaml (hunk 1 above)
Comment (nikhiljindal) [inline federation/manifests/federation-apiserver-deployment.yaml]: What does not work if we do not set this?
Reading the documentation, I get the impression that if its a public IP, we can set it to external-hostname.
And if its a cluster IP, we can set it to advertise-address.

Comment (colhom) [inline federation/manifests/federation-apiserver-deployment.yaml]: IIRC, it's something to do with the automatic certificate generation logic and AWS loadbalancers giving your DNS names instead of IP addresses.

File: federation/cluster/common.sh (hunk 2 above)
Comment (colhom) [inline federation/cluster/common.sh]: kubectl will fail because the DNS name for the ELB could not be resolved. I guess we could soak that up by retrying kubectl for a few minutes, but this seems more explicit.

File: federation/cluster/common.sh (hunk 3 above)
Comment (colhom) [inline federation/cluster/common.sh]: GCE (for instance) attaches a Public IP directly to the loadbalancer, so there's nothing to poll as we already have the answer.

**Comment by colhom** [comment]
@quinton-hoole rebased, I'll let you know how it goes.

File: federation/manifests/federation-apiserver-deployment.yaml (hunk 1 above)
Comment (nikhiljindal) [inline federation/manifests/federation-apiserver-deployment.yaml]: IIUC, the DNS name is cluster internal only? Doesnt seem right to set it as external-hostname.

cc @lavalamp

File: federation/cluster/common.sh (hunk 3 above)
Comment (nikhiljindal) [inline federation/cluster/common.sh]: Yes. We dont need to run nslookup for GCE, but I would like to to remove that if condition if we can. Nothing wrong in running nslookup for GCE IP address?

**Comment by quinton-hoole** [comment]
@colhom How do your results look after the rebase?

File: cluster/aws/util.sh [hunk 4]
```diff
@@ -1491,7 +1491,7 @@ function test-build-release {
 # Assumed vars:
 #   Variables from config.sh
 function test-setup {
-  "${KUBE_ROOT}/cluster/kube-up.sh"
+  . "${KUBE_ROOT}/cluster/kube-up.sh"
```
Comment (quinton-hoole) [inline cluster/aws/util.sh]: Is this change really necessary?

Comment (colhom) [inline cluster/aws/util.sh]: Without it, subsequent calls to things like `detect-security-group` will fail as `VPC_ID` is set within `kube-up` and hence will not be availabile if `kube-up.sh` executes in a sub-process.

**Comment by nikhiljindal** [comment]
Our e2e tests on jenkins are failing right now and I suspect this is the culprit.
The cluster fails to come up with the following logs:

```
attempting to get federation-apiserver loadbalancer hostname (1 / 30)
attempting to get federation-apiserver loadbalancer hostname (2 / 30)
attempting to get federation-apiserver loadbalancer hostname (3 / 30)
attempting to get federation-apiserver loadbalancer hostname (4 / 30)
attempting to get federation-apiserver loadbalancer hostname (5 / 30)
attempting to get federation-apiserver loadbalancer hostname (6 / 30)
attempting to get federation-apiserver loadbalancer hostname (7 / 30)
attempting to get federation-apiserver loadbalancer hostname (8 / 30)
attempting to get federation-apiserver loadbalancer hostname (9 / 30)
2016/06/24 09:15:29 e2e.go:218: Error running up: exit status 1
2016/06/24 09:15:29 e2e.go:214: Step 'up' finished in 14m7.817992497s
2016/06/24 09:15:29 e2e.go:114: Error starting e2e cluster. Aborting.
```

cc @kubernetes/sig-cluster-federation

**Comment by nikhiljindal** [comment]
Verified that the test turned red right after this PR was merged.
Reverting
fedred

**Comment by quinton-hoole** [comment]
@colhom Have you managed to figure out what went wrong here?
@nikhiljindal FYI

**Comment by quinton-hoole** [comment]
Sorry, just saw your comments on #28030. We'll have to fix the nslookup issue.
